presidio-image-redactor/presidio_image_redactor/image_processing_engine.py

```diff
@@ -0,0 +1,361 @@
+        contrast, _ = self._get_image_contrast(image)
+
+        c = (
+            self.c_low_contrast
+            if contrast <= self.contrast_threshold
+            else self.c_high_contrast
+        )
+
+        if background_color < self.bg_threshold:
+            adaptive_threshold_image = cv2.adaptiveThreshold(
+                image,
+                255,
+                cv2.ADAPTIVE_THRESH_MEAN_C,
+                cv2.THRESH_BINARY_INV,
+                self.block_size,
+                -c,
+            )
+        else:
+            adaptive_threshold_image = cv2.adaptiveThreshold(
+                image,
+                255,
+                cv2.ADAPTIVE_THRESH_MEAN_C,
+                cv2.THRESH_BINARY,
+                self.block_size,
+                c,
+            )
+
+        metadata = {"C": c, "background_color": background_color, "contrast": contrast}
+        return Image.fromarray(adaptive_threshold_image), metadata
+
+
+class ImageRescaling(ImagePreprocessor):
+    """ImageRescaling class. Rescales images based on their size."""
+
+    def __init__(
+        self,
+        small_size: int = 1048576,
+        large_size: int = 4000000,
+        factor=2,
+        interpolation=cv2.INTER_AREA,
+    ):
+        """Initialize the ImageRescaling class.
+
+        :param small_size: Threshold for small image size.
+        :param large_size: Threshold for large image size.
+        :param factor: Scaling factor for resizing.
+        :param interpolation: Interpolation method for resizing.
+        """
+
+        self.small_size = small_size
+        self.large_size = large_size
+        self.factor = factor
+        self.interpolation = interpolation
+
+    def preprocess_image(
+        self, image: PngImagePlugin.PngImageFile
+    ) -> Tuple[Image.Image, dict]:
+        """Preprocess the image to be analyzed.
+
+        :param image: Loaded PIL image.
+
+        :return: The processed image and metadata (scale_factor).
+        """
+
+        scale_factor = 1
+        if image.size < self.small_size:
+            scale_factor = self.factor
+        elif image.size > self.large_size:
+            scale_factor = 1 / self.factor
+
+        width = int(image.shape[1] * scale_factor)
+        height = int(image.shape[0] * scale_factor)
+        dimensions = (width, height)
+
+        # resize image
+        rescaled_image = cv2.resize(image, dimensions, interpolation=self.interpolation)
+        metadata = {"scale_factor": scale_factor}
+        return Image.fromarray(rescaled_image), metadata
+
+
+class ContrastSegmentedImageEnhancer(ImagePreprocessor):
+    """Class containing all logic to perform contrastive segmentation."""
+
+    def __init__(
+        self,
+        bilateral_filter: Optional[BilateralFilter] = None,
+        adaptive_threshold: Optional[SegmentedAdaptiveThreshold] = None,
+        image_rescaling: Optional[ImageRescaling] = None,
+        low_contrast_threshold: int = 40,
+    ) -> None:
```

Comment:
For consistency, please either use `-> None` or don't for all your no-return methods in this file.